src/syngen/ml/vae/models/features.py

```diff
@@ -419,112 +420,84 @@ def transform(self, data: pd.DataFrame) -> np.ndarray:
         # return data_gen
         return K.one_hot(K.cast(data_gen, "int32"), self.vocab_size)
 
-    def top_k_top_p_filtering(
-        self,
-        logits: np.ndarray,
-        top_k: int = 0,
-        top_p: float = 0,
-        filter_value: int = -1e8,
+    @staticmethod
+    def top_p_filtering(
+            logits: np.ndarray,
+            top_p: float = 0.9
     ):
-        """
-        Filter a distribution of logits using top-k and/or nucleus (top-p) filtering
-        https://gist.github.com/thomwolf/1a5a29f6962089e871b94cbd09daf317
-        https://github.com/ari-holtzman/degen
-        Args:
-            logits: logits distribution shape (..., vocabulary size)
-            top_k: >0 keep only top k tokens with the highest probability (top-k filtering).
-            top_p: >0.0 keep the top tokens with cumulative probability >= top_p
-            (nucleus filtering).
-            filter_value: value to replace small values with.
-        """
+        # Convert logits to TensorFlow tensor
+        logits = tf.convert_to_tensor(logits, dtype=tf.float32)
+
+        # Sort logits and get sorted indices
+        sorted_logits = tf.sort(logits, direction="DESCENDING", axis=-1)
+        sorted_indices = tf.argsort(logits, direction="DESCENDING", axis=-1)
+
+        # Calculate cumulative probabilities
+        cumulative_probs = tf.cumsum(sorted_logits, axis=-1)
+
+        # Remove tokens with cumulative probability above the threshold
+        sorted_indices_to_remove = cumulative_probs >= top_p
+
+        # Shift the indices to the right to keep also the first token above the threshold
+        zeros_for_shift = tf.zeros_like(sorted_indices_to_remove[:, :, :1], dtype=tf.bool)
+        sorted_indices_to_remove = tf.concat([zeros_for_shift, sorted_indices_to_remove[:, :, :-1]], axis=-1)
+
+        # Create a mask for indices to remove
+        batch_size, seq_length, vocab_size = logits.shape
+        update_indices = tf.stack([tf.repeat(tf.range(batch_size), seq_length * vocab_size),
```

Comment:
Maybe it makes sense to decompose this expression?

Reply:
Check pls